sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadViaSDF.java

```diff
@@ -0,0 +1,742 @@
+            != null) {
+          startOffset =
+              kafkaSourceDescription.getInt64(KafkaSourceDescriptionSchemas.START_READ_OFFSET);
+        } else if (kafkaSourceDescription.getInt64(KafkaSourceDescriptionSchemas.START_READ_TIME)
+            != null) {
+          startOffset =
+              consumerSpEL.offsetForTime(
+                  offsetConsumer,
+                  topicPartition,
+                  Instant.ofEpochMilli(kafkaSourceDescription.getInt64("start_read_time")));
+        } else {
+          startOffset = offsetConsumer.position(topicPartition);
+        }
+        return new OffsetRange(startOffset, Long.MAX_VALUE);
+      }
+    }
+
+    @GetInitialWatermarkEstimatorState
+    public Instant getInitialWatermarkEstimatorState(@Timestamp Instant currentElementTimestamp) {
+      return currentElementTimestamp;
+    }
+
+    @NewWatermarkEstimator
+    public MonotonicallyIncreasing newWatermarkEstimator(
+        @WatermarkEstimatorState Instant watermarkEstimatorState) {
+      return new MonotonicallyIncreasing(watermarkEstimatorState);
+    }
+
+    @GetSize
+    public double getSize(@Element Row kafkaSourceDescription, @Restriction OffsetRange offsetRange)
+        throws Exception {
+      Map<String, Object> updatedConsumerConfig =
+          overrideBootstrapServersConfig(consumerConfig, kafkaSourceDescription);
+      double numOfRecords = 0.0;
+      if (offsetRange.getTo() != Long.MAX_VALUE) {
+        numOfRecords = (new OffsetRangeTracker(offsetRange)).getProgress().getWorkRemaining();
+      } else {
+        TopicPartition topicPartition =
+            new TopicPartition(
+                kafkaSourceDescription.getString(KafkaSourceDescriptionSchemas.TOPIC),
+                kafkaSourceDescription.getInt32(KafkaSourceDescriptionSchemas.PARTITION));
+        KafkaLatestOffsetEstimator offsetEstimator =
+            new KafkaLatestOffsetEstimator(
+                consumerFactoryFn.apply(
+                    KafkaIOUtils.getOffsetConsumerConfig(
+                        "size-" + topicPartition.toString(),
+                        offsetConsumerConfig,
+                        updatedConsumerConfig)),
+                topicPartition);
+        numOfRecords =
+            (new GrowableOffsetRangeTracker(offsetRange.getFrom(), offsetEstimator))
+                .getProgress()
+                .getWorkRemaining();
+      }
+
+      // Before processing elements, we don't have a good estimated size of records and offset gap.
+      if (avgOffsetGap != null) {
+        numOfRecords = numOfRecords / (1 + avgOffsetGap.get());
+      }
+      return (avgRecordSize == null ? 1 : avgRecordSize.get()) * numOfRecords;
+    }
+
+    @SplitRestriction
+    public void splitRestriction(
+        @Element Row kafkaSourceDescription,
+        @Restriction OffsetRange offsetRange,
+        OutputReceiver<OffsetRange> receiver)
+        throws Exception {
+      receiver.output(offsetRange);
```

Comment:
I believe this is the default implementation as well but I don't see it in the documentation so not sure if this is needed or not.

Reply:
Yep the default implementation is doing nothing: https://github.com/apache/beam/blob/591de3473144de54beef0932131025e2a4d8504b/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/ByteBuddyDoFnInvokerFactory.java#L300-L307

But I still want to keep the implementation there to make it explicitly.

Reply:
I think we should remove it and update the DoFn documentation stating the default clearly.